### PR TITLE
Melodic devel general spawn multiple robots

### PR DESCRIPTION
**Pull request by ardalantj**
1. Added proper namespacing and prefixing for running quad software for multiple robots
2. Cleaned up and reorganized launch files

### CODE REVIEW

File: quad_utils/rviz/example_with_terrain.rviz [hunk 1]
```diff
@@ -454,6 +456,6 @@ Window Geometry:
     collapsed: false
   Views:
     collapsed: true
-  Width: 1440
-  X: 1920
-  Y: 0
+  Width: 1613
+  X: 67
+  Y: 27
```
Comment (jcnorby) [inline quad_utils/rviz/example_with_terrain.rviz:461]: Revert all these changes except switching the state topic name/tf prefix

File: quad_utils/src/rviz_interface.cpp [hunk 2]
```diff
@@ -68,18 +69,18 @@ RVizInterface::RVizInterface(ros::NodeHandle nh) {
...
+  nh.param<std::vector<int>>("rviz_interface/colors/back_left",
+                             back_left_color_, {0, 0, 255});
+  nh.param<std::vector<int>>("rviz_interface/colors/front_right",
+                             front_right_color_, {0, 255, 0});
+  nh.param<std::vector<int>>("rviz_interface/colors/back_right",
+                             back_right_color_, {0, 0, 255});
+  nh.param<std::vector<int>>("rviz_interface/colors/net_grf", net_grf_color_,
+                             {255, 0, 0});
+  nh.param<std::vector<int>>("rviz_interface/colors/individual_grf",
+                             individual_grf_color_, {255, 0, 0});
```
Comment (jcnorby) [inline quad_utils/src/rviz_interface.cpp:83]: I think we want "> >" as ">>" is an existing operator.

File: quad_utils/src/rviz_interface.cpp [hunk 3]
```diff
@@ -531,7 +532,7 @@ void RVizInterface::robotStateCallback(
     state_estimate_trace_pub_.publish(state_estimate_trace_msg_);
 
   } else if (pub_id == GROUND_TRUTH) {
-    transformStamped.child_frame_id = "/ground_truth/body";
+    transformStamped.child_frame_id = tf_prefix_ + "/body";
```
Comment (jcnorby) [inline quad_utils/src/rviz_interface.cpp:535]: Is there a reason why this one just has tf_prefix_ + "/body" whereas the others have estimate and trajectory? I think we would want to leave "ground_truth" in there.

File: quad_utils/rviz/example_with_terrain.rviz (hunk 1 above)
Comment (ardalantj) [reply quad_utils/rviz/example_with_terrain.rviz:461]: sounds good

File: quad_simulator/gazebo_scripts/src/controller_plugin.cpp [hunk 4]
```diff
@@ -88,15 +88,20 @@ bool SpiritController::init(hardware_interface::EffortJointInterface* hw,
   quad_utils::loadROSParam(n, "/topics/control/joint_command",
                            joint_command_topic);
 
+  auto ns = n.getNamespace();
```
[the remark this answers is not in the document]
Comment (ardalantj) [reply quad_simulator/gazebo_scripts/src/controller_plugin.cpp:91]: done

File: quad_simulator/spirit_description/sdf_mesh/spirit.sdf [hunk 5]
```diff
@@ -1502,7 +1502,7 @@ SOFTWARE.
     </joint>
     <!-- Setup ros control -->
     <plugin name='gazebo_ros_control' filename='libgazebo_ros_control.so'>
-      <robotNamespace>/spirit</robotNamespace>
+      <!-- <robotNamespace></robotNamespace> -->
```
[the remark this answers is not in the document]
Comment (ardalantj) [reply quad_simulator/spirit_description/sdf_mesh/spirit.sdf:1505]: done

File: quad_utils/launch/mapping.launch [hunk 6]
```diff
@@ -1,6 +1,6 @@
 <launch>
   
-    <include file="$(find quad_utils)/launch/load_params.launch" />
+    <!-- <include file="$(find quad_utils)/launch/load_params.launch" /> -->
```
[the remark this answers is not in the document]
Comment (ardalantj) [reply quad_utils/launch/mapping.launch:3]: removed

File: quad_utils/src/rviz_interface.cpp (hunk 3 above)
Comment (ardalantj) [reply quad_utils/src/rviz_interface.cpp:535]: The visualization.launch has a ground_truth namespace already.

File: quad_utils/launch/mapping.launch [hunk 7]
```diff
@@ -37,9 +38,9 @@
     <!-- Launch the grid map filters demo node. -->
     <node pkg="grid_map_demos" type="filters_demo" name="grid_map_filter_node">
         <!-- Input topic -->
-        <param name="input_topic" value="/terrain_map_raw" />
+        <param name="input_topic" value="/$(arg namespace)/terrain_map_raw" />
         <!-- Output topic -->
-        <param name="output_topic" value="/terrain_map" />
+        <param name="output_topic" value="/$(arg namespace)/terrain_map" />
         <!-- Load grid map filter chain configuration -->
         <rosparam command="load" file="$(find quad_utils)/config/filter_chain.yaml" />
```
[the remark this answers is not in the document]
Comment (ardalantj) [reply quad_utils/launch/mapping.launch:45]: done

**Comment by ardalantj** [comment]
@jcnorby FYI, the load params requires some more changes that I prefer doing in a separate PR.

File: external/grid_map_pcl/src/helpers.cpp [hunk 8]
```diff
@@ -7,92 +7,102 @@
...
+#include "grid_map_pcl/GridMapPclLoader.hpp"
+#include "grid_map_ros/GridMapRosConverter.hpp"
 
-namespace grid_map{
+namespace grid_map {
 namespace grid_map_pcl {
 
-
-void setVerbosityLevelToDebugIfFlagSet(const ros::NodeHandle& nh) {
+void setVerbosityLevelToDebugIfFlagSet(const ros::NodeHandle &nh) {
```
[the remark this answers is not in the document]
Comment (ardalantj) [reply external/grid_map_pcl/src/helpers.cpp:29]: No there is something up with my vscode formatting, will fix it

File: quad_utils/launch/quad_gazebo.launch [hunk 9]
```diff
@@ -25,80 +25,72 @@ SOFTWARE.
 -->
```
[the remark this answers is not in the document]
Comment (ardalantj) [reply quad_utils/launch/quad_gazebo.launch:25]: The extra multi robot launch has been removed. The last step would be to try and run on hardware to sort of any namespacing additions needed

**Comment by jcnorby** [comment]
Okay, here's a list of features that don't quite work as expected yet. Let me know if you want me to break these into individual comments. Once these are working I'll give a code/style review:
- planning.launch should not run on sim launch
- Global and local plan visualizations are gone
- Ground truth state trace is turned on in rviz, should be off by default
- visualization.launch on its own dies with a bunch of error messages
- MATLAB bag parsing script processLog.m only shows one available topic, /terrain_map, and I don't think wouldve worked even if those other topics were there (need to add namespace to hardcoded topic names)
- Clicked point does not cause global planner to replan (global planner is probably looking for /robot_1/clicked_point, not /clicked_point which is published by rviz)

File: quad_utils/src/rviz_interface.cpp (hunk 3 above)
Comment (jcnorby) [reply quad_utils/src/rviz_interface.cpp]: This has been resolved in the most recent push.

**Comment by jcnorby** [comment]
multi_robot_plan.launch copies most of the code that is happening in planning.launch, lets simplify so multi_robot_plan.launch directly calls planning.launch twice with different namespace arguments (similar to how quad_gazebo.launch calls robot.launch twice).

File: nmpc_controller/src/nmpc_controller.cpp [hunk 10]
```diff
@@ -1,19 +1,42 @@
 #include "nmpc_controller/nmpc_controller.h"
 
-NMPCController::NMPCController() {
+NMPCController::NMPCController(ros::NodeHandle nh, int robot_id) {
```
Comment (ardalantj) [inline nmpc_controller/src/nmpc_controller.cpp:3]: missing & for nh

File: quad_simulator/a1_description/package.xml [hunk 11]
```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0"?>
+<package format="2">
+    <name>a1_description</name>
+    <version>0.0.0</version>
+    <description>The a1_description package</description>
+
+    <maintainer email="jimingre@andrew.cmu.edu">unitree</maintainer>
+    <license>TODO</license>
```
Comment (ardalantj) [inline quad_simulator/a1_description/package.xml:8]: Is this supposed to be left as TODO?

File: robot_driver/src/robot_driver.cpp [hunk 12]
```diff
@@ -136,7 +144,7 @@ RobotDriver::RobotDriver(ros::NodeHandle nh, int argc, char** argv) {
   last_state_time_ = std::numeric_limits<double>::max();
 
   // Set joint torque limits
-  torque_limits_ << 21, 21, 32;
+  // torque_limits_ << 21, 21, 32;
```
Comment (ardalantj) [inline robot_driver/src/robot_driver.cpp:147]: remove?

File: robot_driver/src/robot_driver.cpp [hunk 13]
```diff
@@ -1,10 +1,14 @@
...
 
-RobotDriver::RobotDriver(ros::NodeHandle nh, int argc, char** argv) {
+RobotDriver::RobotDriver(ros::NodeHandle nh, int argc, char **argv) {
   nh_ = nh;
   argc_ = argc;
   argv_ = argv;
 
+  // Load system parameters from launch file (not in config file)
+  nh.param<std::string>("robot_type", robot_name_, "spirit");
+  // nh.param<std::string>("ns", ns_, "/");
```
Comment (ardalantj) [inline robot_driver/src/robot_driver.cpp:10]: remove?

Comment (jrenaf) [reply robot_driver/src/robot_driver.cpp]: Changed to loadRosParam

File: robot_driver/src/robot_driver.cpp (hunk 12 above)
Comment (jrenaf) [reply robot_driver/src/robot_driver.cpp]: Deleted.

File: quad_simulator/a1_description/package.xml (hunk 11 above)
Comment (jrenaf) [reply quad_simulator/a1_description/package.xml:8]: Leave to license pull request.

**Comment by jcnorby** [comment]
Okay, second round of functionality checks:
- Default configuration should be only launching one robot (although we can swap this right before merging)
- Local plan position trajectories are gone
- Logging scripts still broken - they need to be aware of the namespace and subscribe to the topics accordingly (see next comment), and processLog.m/parseQuadBag.m likely need namespace variables so they can read the topics.
- Multi_robot_plan.launch copy-pastes most of the code within planning.launch twice, but many of the included files or nodes in planning.launch aren't being passed the namespace (e.g. logging.launch). Should modify multi_robot_plan.launch to just call planning.launch twice with appropriate namespacing
- visualization.launch crashes if launched on its own. This was the previous workflow to inspect bags or screen capture them - either we need to fix it or introduce a new workflow/launch file which can be run to easily view bags

**Pull request closed without merging**